src/Microsoft.ApiDesignGuidelines.Analyzers/Core/IdentifiersShouldNotMatchKeywords.cs

```diff
@@ -1,69 +1,621 @@
-                                                                             DiagnosticCategory.Naming,
-                                                                             DiagnosticSeverity.Warning,
-                                                                             isEnabledByDefault: false,
+                                                                             Category,
+                                                                             Severity,
+                                                                             isEnabledByDefault: IsEnabledByDefault,
                                                                              description: s_localizableDescription,
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
-                                                                             customTags: WellKnownDiagnosticTags.Telemetry);
+                                                                             helpLinkUri: HelpLinkUri,
+                                                                             customTags: CustomTags);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(MemberParameterRule, MemberRule, TypeRule, NamespaceRule);
 
         public override void Initialize(AnalysisContext analysisContext)
         {
-            
+            analysisContext.RegisterSymbolAction(AnalyzeMemberRule,
+                SymbolKind.Event, SymbolKind.Method, SymbolKind.Property);
+
+            analysisContext.RegisterSymbolAction(AnalyzeMemberParameterRule,
+                SymbolKind.Method);
+        }
+
+        private void AnalyzeMemberRule(SymbolAnalysisContext context)
+        {
+            ISymbol symbol = context.Symbol;
+            if (symbol.GetResultantVisibility() != SymbolVisibility.Public)
+            {
+                return;
+            }
+
+            string matchingKeyword;
+            if (!IsKeyword(symbol.Name, out matchingKeyword))
+            {
+                return;
+            }
+
+            // IsAbstract returns true for both abstract class members and interface members.
+            if (symbol.IsVirtual || symbol.IsAbstract)
+            {
+                context.ReportDiagnostic(
+                    symbol.CreateDiagnostic(
+                        MemberRule,
+                        FormatSymbolName(symbol),
+                        matchingKeyword));
+            }
+        }
+
+        private void AnalyzeMemberParameterRule(SymbolAnalysisContext context)
+        {
+            var method = (IMethodSymbol)context.Symbol;
+            if (method.GetResultantVisibility() != SymbolVisibility.Public)
+            {
+                return;
+            }
+
+            // IsAbstract returns true for both abstract class members and interface members.
+            if (!method.IsVirtual && !method.IsAbstract)
```

Comment:
Actually, the beginning of AnalyzeMemberRule and AnalyzeMemberParameterRule is identical. Couldn't these be combined into one for less repetition and potentially better perf?